srk/src/nestedLoops.ml

```diff
@@ -0,0 +1,311 @@
+
+
+(** Decide if v is a header for the i-th SCC in wg by looking at whether
+    its coming edges are all in this scc.
+    `f` is an SCC function that maps each v to its SCC component number.
+*)
+let is_header wg f i_th_scc v = 
+    fold_pred_e 
+        (fun (u, weight, v) (acc) -> (f u <> i_th_scc) || acc)
+        wg
+        v
+        false
+    
+(** find a header and a backedge that goes into the header *)
+(* replace filter with find *)
+let find_header_backedge wg f i_th_scc scc =
+    printf "... finding a header and a backedge ... \n";
+    let headers = List.filter (is_header wg f i_th_scc) scc in
+    match headers with
+    | [] -> assert false
+    (* we assume that only allowing 1 header is enough *)
+    | hd :: _ -> 
+        let backedge_list = 
+        fold_pred_e 
+            (fun (u, weight, v) l -> if (List.mem u scc) then (u, weight) :: l else l)
+            wg
+            hd
+            []
+        in
+        match backedge_list with
+        | [] -> assert false
+        | t :: _ -> (fst t, hd, snd t)
+
+(** replace the pos-th element in l with a *)
+let replace l pos a = 
+    List.mapi (fun i x -> if i = pos then a else x) l
+
+(** Assign child loops to parent loops, given two lists of loops.
+    `f` is the SCC function that maps vertex to SCC component number, which a by-product of
+    parent loop computation.
+ *)
+let rec assign_child_loops l_parent l_children f =
+    printf "... trying to assign child loops to parent loops\n";
+    printf "parent list:\n";
+    List.iteri (fun i l -> match l with None -> printf "%d:{}\n" i | Some ll -> printf "%d:" i; print_loop ll) l_parent;
+    printf "children list:\n";
+    List.iter (fun l -> print_loop l) l_children;
+
+    match l_children with
+    | [] -> l_parent
+    | hd :: tail -> 
+        let parent_ord = f hd.header in
+        let parent_loop = (List.nth l_parent parent_ord) in
+        match parent_loop with 
+            | None -> assert false 
+            | Some parent_loop ->
+                printf "this child should be assigned to %d -th parent\n" parent_ord;
+                replace 
+                    (assign_child_loops l_parent tail f) 
+                    parent_ord 
+                    (Some { parent_loop with children = hd :: parent_loop.children })
+                    
+
+(** Remove all backedges (top-level) recorded in a list of loops from wg *)
+let rec remove_all_backedges wg l_loops =
+    match l_loops with
+    | [] -> wg
+    | loop_op :: tail -> 
+        begin
+            match loop_op with
+            | None -> remove_all_backedges wg tail
+            | Some loop ->
+            let (u, v, _) = loop.back_edge in
+            let wgn = remove_edge wg u v in
+                remove_all_backedges wgn tail
+        end
+
+(** compute a tree of nested loops, returns a tuple of
+    (a wg with all backedges deleted, a list of loops at the top level)
+*)
+let rec compute_nested_loops wg d =
```

Comment:
It looks like you're computing a tree level-by-level.  I think it would simplify the algorithm if you computed sub-tree by sub-tree (so if the root has k children, you compute the entire sub-tree for the first child, then second, ... through the kth).